netdata: facilitate building netdata with cloud support

leotaku opened the pull request:
This PR adds a `withCloud` option to the netdata package. When enabled it allows the netdata agent to communicate with the netdata cloud.

Other than that, we unconditionally wrap the `netdata-claim.sh` script with a path to openssl and no longer remove the `sbin` output as it is required by the script.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change

The current netdata package does not support connecting to the netdata cloud.
There has been a packaging request for netdata with cloud support in #91549. 
#92291 was merged in preparation to this PR.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

leotaku commented:
This PR also adds a new file for the custom version of `mosquitto` used by netdata. I was unsure if it makes sense for that package to have a `meta` section, so I have left it out of the initial PR.

leotaku commented:
@jtojnar I really don't want to pester you, but maybe it would be possible for you to review this PR, as you have already approved and merged #92291? Thank you for your hard work on the Nixpkgs project!

jtojnar commented on pkgs/tools/system/netdata/mosquitto.nix:
```diff
@@ -0,0 +1,33 @@
+{ stdenv, fetchFromGitHub, c-ares, cmake, libwebsockets_3_2, openssl }:
```
Could you add some comment how does this differ from mosquitto 1.6.10 from nixpkgs and why it cannot be used?

leotaku replied:
Do you mean add a comment here or in the actual file? 

In any case, the Netdata developers seem to have decided that it is necessary for them to fork Mosquitto. From what I can tell, this was done to accommodate some external callback functionality they require. Trying to build Netdata with the regular version of Mosquitto fails because of missing symbols/functions.

jtojnar replied:
I guess maybe a comment near the `preConfigure` linking the issue should be enough.

leotaku replied:
I have added a comment to the top of the `mosquitto.nix` file as I think that makes the most sense. If you disagree please unresolve this conversation.

jtojnar commented on pkgs/tools/system/netdata/default.nix:
```diff
@@ -57,18 +60,20 @@ in stdenv.mkDerivation rec {
-    "--localstatedir=/var"
-    "--sysconfdir=/etc"
-  ];
+  configureFlags = [ "--localstatedir=/var" "--sysconfdir=/etc" ]
```
Could you keep this on multiple lines?

leotaku replied:
I've tried to fix the formatting in my latest commit.

jtojnar commented on pkgs/tools/system/netdata/default.nix:
```diff
@@ -57,18 +60,20 @@ in stdenv.mkDerivation rec {
+    ++ optional withCloud "--enable-cloud";
 
   postFixup = ''
-    rm -r $out/sbin
```
Why this unrelated removal?

leotaku replied:
> Other than that, we unconditionally wrap the netdata-claim.sh script with a path to openssl and no longer remove the sbin output as it is required by the script.

Maybe it was unclear from the original message, but the included `netdata-claim.sh` script (which is used to register netdata agents to cloud) looks for netdata binaries under `sbin`. I personally thought that just keeping it would be preferable to rewriting the script.

jtojnar replied:
Right, sorry, misread that.

jtojnar commented on pkgs/tools/system/netdata/default.nix:
```diff
@@ -57,18 +60,20 @@ in stdenv.mkDerivation rec {
-    "--sysconfdir=/etc"
-  ];
+  configureFlags = [ "--localstatedir=/var" "--sysconfdir=/etc" ]
+    ++ optional withCloud "--enable-cloud";
```
Again I will ask how does this affect closure size and if it would not make sense to enable it unconditionally.

leotaku replied:
Closure size goes from `148743560` to `149251816`, so it's basically irrelevant. 

I would be very grateful if you could tell me what the policy about adding new configuration options currently is. I generally just see the many e.g. `withCups, withDBengine, withIpmi, withNetfilter, withSsl` configuration options, many of which seem even less relevant than e.g. `withCloud`, and assume that adding new options is good practice. But your comments make me think that they might just be there for backwards compatibility and "no options" is preferred?

jtojnar replied:
I do not think there is a policy. There are few legitimate reasons for having the options overridable (when it makes the closure significantly larger, when it adds an unfree dependency to otherwise free package…) but most of the time it looks to me like a cargo-cult.

I typically ask myself: “Would someone want to disable this feature? If there were an overridable option, would someone bother overriding it? Are the benefits commensurate with the added complexity to the expression?” I would only add an option when I answered “Yes” to all three of the questions.

But in the end, the decision is up to the maintainers of each individual package, since they will bear the complexity.

leotaku replied:
I think `withCloud` could be interesting to some users that would like to ensure that none of their system analytics are streamed to servers owned by the Netdata company. [There seems to be a way](https://github.com/netdata/netdata/issues/9043) to disable the cloud connection at runtime, but it does seem like disabling cloud support at built time is still the only option to ensure no connection is made.

Personally, I think that's useful enough to keep the `withCloud` option, which, in my opinion, also does not add a lot of complexity to the derivation.

wmertens replied:
Looks like the cloud disabling is now fixed upstream?

Given that, I propose the package just gets built with cloud support.

jtojnar commented on pkgs/tools/system/netdata/default.nix:
```diff
@@ -57,18 +60,20 @@ in stdenv.mkDerivation rec {
-  preConfigure = optionalString (!stdenv.isDarwin) ''
+  preConfigure = optionalString withCloud ''
+    mkdir -p externaldeps
+    cp -r "${mosquitto}/lib" externaldeps/mosquitto;
+    cp -r "${libwebsockets_3_2}/lib" externaldeps/libwebsockets;
```
Ugh https://github.com/netdata/netdata/blob/3d33eea007967ffe70f7918e6b617366cb34dd28/CMakeLists.txt#L916-L925

Does not someone (some other distro, pull request on the repo) have a patch that would allow using shared library from the system?

leotaku replied:
I noticed that this is terrible practice by the upstream project, but I chose to simply adapt the build script for the initial PR.

jtojnar replied:
Checking out [other distros](https://repology.org/project/netdata/versions) (Alpine, Arch, Debian, Fedora, [Gentoo](https://gitweb.gentoo.org/repo/gentoo.git/tree/net-analyzer/netdata/netdata-1.23.2.ebuild?id=2eb0beba59cc05194c330ad0178a589f1605e902#n98)) I found a relevant issue:

https://github.com/netdata/netdata/issues/8961

leotaku replied:
At least some of the developers think that accommodating Linux distributions is generally a good idea, however I am not that confident that we will see actual changes anytime soon. In any case, I have subscribed to the Issue and will of course update the PR if I notice any changes.

leotaku replied:
[According to one of the developers](https://github.com/netdata/netdata/issues/8961#issuecomment-762337210), Netdata will soon be using a new implementation of the [Agent-Cloud-Link](https://github.com/netdata/netdata/pull/10315) which would only require unpatched and common dependencies (JSON-C and OpenSSL). I would be more than happy to modify this PR when this change lands, or if it could be agreed to merge this PR in its current "hackish" state, open a new PR to then fix these issues.

underhood replied:
@leotaku hi, you are right I am pushing for the new implementation of the link as you mentioned. Called `ACLK-NG`, started originally as a skunkworks project it is starting to take shape and we want to merge it ASAP.
While it works already it will take some time to weatherproof it and catch all possible bugs. This is the motive why originally it will be released as a fallback (in case ACLK-Legacy cannot be built e.g. due to missing custom mosquito lib), with the possibility to enforce its use by passing `--aclk-ng` to the installer.

The main motivation to do this for me was to enable full-featured Netdata in every distribution package manager (e.g. remove the need for bundling libraries and other "hacky" things). As a side effect, it also seems to perform better (throughput doesn't drop with latency as was the case with Legacy).

leotaku committed:
Reformat configureFlags list to span multiple lines

rissson reviewed (changes requested):
When testing this with Netdata 1.25.0 (basically applying your changes on top of `nixos-20.09`), it fails with
```
/nix/store/hiwz81i1g3fn3p0acjs042a4h5fri6dh-binutils-2.31.1/bin/ld: externaldeps/libwebsockets/libwebsockets.a(unix-plugins.c.o): undefined reference to symbol 'dlclose@@GLIBC_2.2.5'
  /nix/store/hiwz81i1g3fn3p0acjs042a4h5fri6dh-binutils-2.31.1/bin/ld: /nix/store/bdf8iipzya03h2amgfncqpclf6bmy3a1-glibc-2.32/lib/libdl.so.2: error adding symbols: DSO missing from command line
```
However, it works fine when building from this commit https://github.com/netdata/netdata/commit/593e1b6dbc6c979955e503d431d32c0dce1a2e09 and not using the bundled `libwebsockets`. I think we can wait for the next Netdata release to be able to use this.

leotaku commented on pkgs/tools/system/netdata/mosquitto.nix:
```diff
@@ -0,0 +1,33 @@
+
+  postInstall = ''
+    cp $out/lib/libmosquitto_static.a $out/lib/libmosquitto.a
+    cp $out/include/mosquitto.h $out/lib/mosquitto.h
```
I don't fully understand what your change here is supposed to do. With this change, the source and target path are identical.

rissson replied:
Actually this was right, I probably was a bit too tired when I reviewed this, sorry.

leotaku commented on pkgs/tools/system/netdata/default.nix:
```diff
@@ -57,18 +60,22 @@ in stdenv.mkDerivation rec {
+  preConfigure = optionalString withCloud ''
+    mkdir -p externaldeps
+    cp -r "${mosquitto}/lib" externaldeps/mosquitto;
+    cp -r "${libwebsockets_3_2}/lib" externaldeps/libwebsockets;
```
I also think would be a good idea to wait for the issues with packaging Netdata to be fixed upstream before we merge this PR.

rissson replied:
Let's just wait for `libwebsockets` though. If I understand correctly, it's not likely we will be able to use upstream `mosquitto`.

leotaku replied:
Sure, that makes sense. Also, I think they are making the build process more configurable, which means we probably won't have to move files to fixed locations anymore.

rissson replied:
I opened https://github.com/NixOS/nixpkgs/pull/107306 updating netdata to the latest version. I am currently testing your PR on top of it.

rissson replied:
It seems to work :tada:
I'll push my changes somewhere tomorrow and link them here.

rissson replied:
https://gitlab.com/lama-corp/infra/dotshabka/-/commit/f57284ac1851944d1d765c9b323f95cb1d84c51b

leotaku replied:
I suppose the best way to incorporate your changes would be to merge them in from your GitLab repository. I assume this means that you would not be added as a co-author of this PR, despite having contributed code. Would you be fine with that?

rissson replied:
Well the actual work I put in this is in #107306, I just applied the changes from this PR to it. However, if you want to add someone as a co-author, you can always add a tag at the end of your commit message like this:
```
Co-authored-by: Name <email>
```

leotaku replied:
I will use the `Co-authored-by: Name <email>` tag so that I can also include your update to mosquitto.

SuperSandro2000 commented on pkgs/tools/system/netdata/default.nix:
```diff
@@ -1,18 +1,20 @@
 , withIpmi ? (!stdenv.isDarwin), freeipmi
 , withNetfilter ? (!stdenv.isDarwin), libmnl, libnetfilter_acct
 , withSsl ? true, openssl
+, withCloud ? true, libwebsockets_3_2, json_c
```
The cloud is unfree and should probably not be enabled when the user does not allow unfree packages.

rissson replied:
If I understand correctly the agent-side of the cloud code is released under GPLv3, whereas the server part of Netdata Cloud isn't released at all.

leotaku replied:
I can confirm that the client-side Netdata code is 100% Open Source while the server-side is not. Nixpkgs already contains other packages with this same arrangement (hercules-ci-agent, cachix and the GitHub CLI, just from the top of my head) and they are generally considered "free".

SuperSandro2000 commented:
Result of `nixpkgs-review pr 92782` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>netdata</li>
  </ul>
</details>

SuperSandro2000 commented on pkgs/tools/system/netdata/mosquitto.nix:
```diff
@@ -0,0 +1,33 @@
+  '';
+
+  postInstall = ''
+    cp $out/lib/libmosquitto_static.a $out/lib/libmosquitto.a
```
Can we use ln here?

leotaku replied:
Sorry, but I am not quite sure what you mean by "using ln", could you maybe elaborate?

rissson replied:
I think they mean doing something like this:
```
ln -s $out/lib/libmosquitto_static.a $out/lib/libmosquitto.a
```

leotaku replied:
Apologies, `I` and `l` look the same to me in my system variable-width font, so I got confused. It does not seem possible to use symbolic links here (build fails) but hard links seem to be fine.

leotaku replied:
I'm not sure if hard linking is something that is commonly done in Nixpkgs and whether it has any particular advantages or disadvantages. A very naive grep-based investigation shows that copying files is done much more frequently than hardlinking, so I'd probably just keep the `cp`.

```shell
$ cd nixpkgs
$ rg "cp[ ][^(-r)]" | wc -l
1451
$ rg "ln[ ][^(-s)]" | wc -l
29
```

SuperSandro2000 replied:
cp is useful when moving files from source to out but ln should be used when symlinking static libs in out or even mv.

leotaku replied:
Ok, I have switched the two lines to hardlinking now.

> ln should be used when symlinking static libs in out or even mv

You are aware that we are not using symlinking here, right? I am now using hardlinking, which is a different concept which is why I am unsure if I should use it in derivations.

SuperSandro2000 commented:
Result of `nixpkgs-review pr 92782` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>netdata</li>
  </ul>
</details>

```
  CC       collectors/proc.plugin/proc_uptime.o
  CC       collectors/proc.plugin/sys_kernel_mm_ksm.o
  CC       collectors/proc.plugin/sys_block_zram.o
  CC       collectors/proc.plugin/sys_devices_system_edac_mc.o
  CC       collectors/proc.plugin/sys_devices_system_node.o
  CC       collectors/proc.plugin/sys_fs_btrfs.o
  CC       collectors/proc.plugin/sys_class_power_supply.o
  CC       collectors/proc.plugin/sys_class_infiniband.o
  CC       collectors/tc.plugin/plugin_tc.o
  GEN      netdata
/nix/store/kcp0y7g1ir4dxq2gqz687i4vb7gqy8s0-binutils-2.31.1/bin/ld: externaldeps/libwebsockets/libwebsockets.a(unix-plugins.c.o): undefined reference to symbol 'dlclose@@GLIBC_2.2.5'
/nix/store/kcp0y7g1ir4dxq2gqz687i4vb7gqy8s0-binutils-2.31.1/bin/ld: /nix/store/1yvpgm763b3hvg8q4fzpzmflr5674x4j-glibc-2.32-10/lib/libdl.so.2: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:4371: netdata] Error 1
make[2]: Leaving directory '/build/source'
make[1]: *** [Makefile:5506: all-recursive] Error 1
make[1]: Leaving directory '/build/source'
make: *** [Makefile:2987: all] Error 2
```

leotaku committed:
Remove redundant semicolons

rissson commented:
@leotaku  With #107306 merged, would you mind rebasing on master?

leotaku committed:
Update libmosquitto for new Netdata version

Co-authored-by: Marc Schmitt <marc.schmitt@risson.space>

leotaku committed:
Update build script for new Netdata version

Co-authored-by: Marc Schmitt <marc.schmitt@risson.space>

leotaku committed:
Add comment to mosquitto.nix file

leotaku committed:
netdata: facilitate building netdata with cloud support

This commit adds a `withCloud` option to the netdata package. When
enabled it allows the netdata agent to communicate with the netdata
cloud.

Other than that, we unconditionally wrap the `netdata-claim.sh` script
with a path to openssl and no longer remove the `sbin` output as it is
required by the script.

SuperSandro2000 commented on pkgs/tools/system/netdata/mosquitto.nix:
```diff
@@ -0,0 +1,39 @@
+    "-DWITH_DOCS=NO"
+  ];
+
+  buildPhase = ''
+    cd lib
+    make install
+  '';
```
Please use preBuild and makeFlags.

leotaku replied:
I hope my solution is approximately what you had in mind.

leotaku committed:
Replace explicit buildPhase with makeFlags

leotaku committed:
Use hardlinks for duplicate files in $out

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review which is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 92782` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>netdata</li>
</ul>
</details>

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review which is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 92782` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>netdata</li>
</ul>
</details>

SuperSandro2000 commented on pkgs/tools/system/netdata/mosquitto.nix:
```diff
@@ -0,0 +1,38 @@
+  postInstall = ''
+    ln $out/lib/libmosquitto_static.a $out/lib/libmosquitto.a
+    ln $out/include/mosquitto.h $out/lib/mosquitto.h
+  '';
```
Not sure if we want to add a meta section here.

leotaku replied:
Sorry for keeping you waiting, I don't think I got any notifications for your comments. Honestly, I think it should be fine not having the meta section, especially since I would like to remove this package from Nixpkgs as soon as [ACLK-NG](https://github.com/NixOS/nixpkgs/pull/92782#discussion_r583834062) is present in the Netdata package provided by Nixpkgs.

SuperSandro2000 replied:
Packages require a meta section.

rissson replied:
I mean, this package will likely disappear in a few weeks, plus it's only used by netdata.

SuperSandro2000 reviewed (approved):
diff LGTM

underhood commented:
Hi guys, ACLK-NG has been merged and is present in current nightlies and will be present in the next Netdata release. Currently is a fallback for when ACLK LEgacy can't be built (e.g. missing dependency) or can be enabled by passing `--aclk-ng` parameter to the installer.

SuperSandro2000 commented:
@leotaku the build is failing

@ofborg build

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 92782` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>netdata</li>
</ul>
</details>

leotaku commented:
The new Netdata release (1.30.0) already has ACLK-NG support. This means that this pull request is no longer relevant.
